src/ncp/ncp_spinel.cpp

```diff
@@ -0,0 +1,292 @@
+    uint8_t           header;
+    otbrError         error = OTBR_ERROR_NONE;
+
+    unpacked = spinel_datatype_unpack(aFrame, aLength, "CiiD", &header, &cmd, &key, &data, &len);
+    VerifyOrExit(unpacked > 0, error = OTBR_ERROR_PARSE);
+    VerifyOrExit(SPINEL_HEADER_GET_TID(header) == 0, error = OTBR_ERROR_PARSE);
+    VerifyOrExit(cmd == SPINEL_CMD_PROP_VALUE_IS);
+    HandleValueIs(key, data, static_cast<uint16_t>(len));
+
+exit:
+    if (error != OTBR_ERROR_NONE)
+    {
+        otbrLogResult(error, "HandleNotification: %s", __FUNCTION__);
+    }
+}
+
+void NcpSpinel::HandleResponse(spinel_tid_t aTid, const uint8_t *aFrame, uint16_t aLength)
+{
+    spinel_prop_key_t key;
+    spinel_size_t     len = 0;
+    spinel_ssize_t    unpacked;
+    uint8_t          *data = nullptr;
+    uint32_t          cmd;
+    uint8_t           header;
+    otError           error          = OT_ERROR_NONE;
+    FailureHandler    failureHandler = nullptr;
+
+    unpacked = spinel_datatype_unpack(aFrame, aLength, "CiiD", &header, &cmd, &key, &data, &len);
+    VerifyOrExit(unpacked > 0, error = OT_ERROR_PARSE);
+
+    VerifyOrExit(cmd == SPINEL_CMD_PROP_VALUE_IS && key == mWaitingKeyTable[aTid], error = OT_ERROR_INVALID_STATE);
+
+    switch (key)
+    {
+    case SPINEL_PROP_NET_ROLE:
+    {
+        spinel_net_role_t spinelRole;
+        failureHandler = [this](otError aError) { CallAndClear(mGetDeviceRoleHandler, aError, mDeviceRole); };
+
+        unpacked = spinel_datatype_unpack(data, len, "i", &spinelRole);
+        VerifyOrExit(unpacked > 0, error = OT_ERROR_PARSE);
+
+        mDeviceRole = SpinelRoleToDeviceRole(spinelRole);
+        CallAndClear(mGetDeviceRoleHandler, error, mDeviceRole);
+        break;
+    }
+    default:
+        break;
+    }
+
+exit:
+    FreeTid(aTid);
+
+    if (error == OT_ERROR_INVALID_STATE)
+    {
+        otbrLogCrit("Received unexpected response with cmd:%u, key:%u, waiting key:%u for tid:%u", cmd, key,
+                    mWaitingKeyTable[aTid], aTid);
+    }
+    else if (error == OT_ERROR_PARSE)
+    {
+        otbrLogCrit("Error parsing response with tid:%u", aTid);
+        if (failureHandler)
+        {
+            failureHandler(error);
+        }
+    }
+}
+
+void NcpSpinel::HandleValueIs(spinel_prop_key_t aKey, const uint8_t *aBuffer, uint16_t aLength)
+{
+    otbrError      error = OTBR_ERROR_NONE;
+    spinel_ssize_t unpacked;
+
+    switch (aKey)
+    {
+    case SPINEL_PROP_LAST_STATUS:
+    {
+        spinel_status_t status = SPINEL_STATUS_OK;
+
+        unpacked = spinel_datatype_unpack(aBuffer, aLength, "i", &status);
+        VerifyOrExit(unpacked > 0, error = OTBR_ERROR_PARSE);
+
+        otbrLogInfo("NCP last status: %s", spinel_status_to_cstr(status));
+        break;
+    }
+    case SPINEL_PROP_NET_ROLE:
+    {
+        spinel_net_role_t role;
+
+        unpacked = spinel_datatype_unpack(aBuffer, aLength, "C", &role);
+        VerifyOrExit(unpacked > 0, error = OTBR_ERROR_PARSE);
```

Comment:
nit: may wrap the function `spinel_datatype_unpack` and let it return an OT error.

Reply:
Updated.